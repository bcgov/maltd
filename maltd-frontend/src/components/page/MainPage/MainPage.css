body {
  color: #303030 !important;
  background-color: #dcdcdc !important;
}

.container {
  max-width: 600px !important;
}

.top-spacing {
  padding-top: 30px;
  display: flex;
}

.less-spacing-top {
  top: -10px;
  width: 50%;
  margin: auto;
  display: flex;
  flex-direction: column;
  background-color: #f2f2f2;
}

@media only screen and (max-width: 700px) {
  /* widen box on smaller screens */
  .less-spacing-top {
    width: 90%;
  }

  /* To stack back button on top of current box */
  #wrapper {
    flex-direction: column;
  }
  .backicon-spacing {
    margin-bottom: 30%;
  }
  .top-spacing {
    padding-top: 15px;
  }
}

/* slightly wider for ipad type screens */
@media only screen and (min-width: 701px) and (max-width: 850px) {
  .less-spacing-top {
    width: 55%;
  }
}
