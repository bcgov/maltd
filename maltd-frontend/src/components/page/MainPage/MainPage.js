import React, { Component } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import "./MainPage.css";
import UserSearch from "../../composite/UserSearch/UserSearch";
import NavBar from "../../base/NavBar/NavBar";
import BackIcon from "../../base/BackIcon/BackIcon";
import UserAccess from "../../composite/UserAccess/UserAccess";

const token = localStorage.getItem("jwt");

export default class MainPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      validInput: false,
      invalidInput: false,
      value: "",
      disabledInput: false,
      disabledButton: true,
      isLoading: false,
      isUserSearch: true,
      projects: [],
      userEmail: null,
      userName: "",
      color: "primary",
      userExists: null,
      items: [],
      selectedDropdownItem: null,
      duplicateErrorMessage: null
    };
  }

  onButtonClick() {
    const { value } = this.state;

    return axios
      .get(`/api/projects`, { headers: { Authorization: `Bearer ${token}` } })
      .then(res => {
        console.log("items res.data", res.data);
        this.setState({
          items: res.data,
          isLoading: true,
          disabledButton: true,
          disabledInput: true
        });
      })
      .then(() => {
        axios.get(`/api/users?q=${value}`, {
          headers: { Authorization: `Bearer ${token}` }
        });
      })
      .then(() => {
        return axios
          .get(`/api/users/${value}`, {
            headers: { Authorization: `Bearer ${token}` }
          })
          .then(result => {
            const { data } = result;
            const finalProjects = [];

            console.log("projects data coming back for user", data.projects);

            data.projects.forEach(proj => {
              let shouldAddProject = false;
              const resources = [];

              proj.resources.forEach(resource => {
                console.log("what is the resource", resource);
                if (resource.status === "member") {
                  resources.push(resource);
                  shouldAddProject = true;
                }
              });

              console.log("proj", proj);
              console.log("resources", resources);

              if (shouldAddProject)
                finalProjects.push({ name: proj.name, resources, id: proj.id });
            });

            console.log("final projects: ", finalProjects);

            this.setState({
              projects: finalProjects,
              isUserSearch: false
            });

            if (data.email) {
              this.setState({ userEmail: data.email });
            }

            if (data.firstName && data.lastName) {
              this.setState({
                userName: `${data.firstName} ${data.lastName}`
              });
            }
          });
      })
      .catch(() => {
        this.clearForm();
      });
  }

  onInputChange(event) {
    this.setState({ userExists: null });
    const val = event.target.value;

    if (val.length === 0) {
      this.setState({
        invalidInput: false,
        validInput: false,
        disabledButton: true,
        color: "primary"
      });
    } else if (val.length < 3) {
      this.setState({
        invalidInput: true,
        color: "danger"
      });
    } else {
      this.setState({
        invalidInput: false,
        validInput: true,
        disabledButton: false,
        color: "primary"
      });
    }

    this.setState({ value: event.target.value });
  }

  onBackClick() {
    this.setState({ isUserSearch: true });

    this.clearForm();

    this.setState({
      userExists: null,
      value: ""
    });
  }

  onKeyEnter(event) {
    const { disabledButton } = this.state;
    if (event.key === "Enter" && !disabledButton) {
      this.onButtonClick();
    }
  }

  removeUserFromProject(projectId) {
    const { value, projects } = this.state;

    return axios
      .delete(`/api/projects/${projectId}/users/${value}`, {
        headers: { Authorization: `Bearer ${token}` }
      })
      .then(() => {
        const updatedProjects = [];
        projects.forEach(proj => {
          if (proj.id !== projectId) {
            updatedProjects.push(proj);
          }
        });
        this.setState({ projects: updatedProjects });
      })
      .catch(() => {});
  }

  addUserToProject() {
    const { selectedDropdownItem, value, projects } = this.state;
    let isDuplicate = false;
    console.log("selecteddropdownitem is", selectedDropdownItem, projects);

    projects.forEach(proj => {
      if (proj.id === selectedDropdownItem.id) {
        isDuplicate = true;
      }
    });

    if (isDuplicate) {
      this.setState({
        duplicateErrorMessage:
          "This project has already been added. Please try again with a different project."
      });
      setTimeout(() => {
        this.setState({ duplicateErrorMessage: null });
      }, 5000);
      return;
    }

    return axios
      .put(`/api/projects/${selectedDropdownItem.id}/users/${value}`, {
        headers: { Authorization: `Bearer ${token}` }
      })
      .then(res => {
        console.log("res or what? ", res);
        const updatedProjects = projects.slice(0);

        updatedProjects.push({
          ...selectedDropdownItem,
          resources: [
            { type: "Dynamics", status: "member" },
            { type: "Sharepoint", status: "member" }
          ]
        });
        this.setState({
          projects: updatedProjects,
          selectedDropdownItem: null,
          duplicateErrorMessage: null
        });
      })
      .catch(() => {});
  }

  updateSelectedDropdownItem(selectedProject) {
    this.setState({ selectedDropdownItem: selectedProject });
  }

  clearForm() {
    this.setState({
      userExists: false,
      isLoading: false,
      disabledButton: true,
      disabledInput: false,
      invalidInput: false,
      validInput: false
    });
  }

  render() {
    const {
      validInput,
      invalidInput,
      value,
      disabledInput,
      disabledButton,
      isLoading,
      isUserSearch,
      projects,
      userEmail,
      userName,
      color,
      userExists,
      items,
      selectedDropdownItem,
      duplicateErrorMessage
    } = this.state;

    const { onLogoutClick } = this.props;

    const inputField = {
      type: "text",
      name: "idir",
      placeholder: "Enter IDIR username to find",
      valid: validInput,
      invalid: invalidInput,
      value,
      disabled: disabledInput
    };

    const generalButton = {
      type: "submit",
      color,
      disabled: disabledButton,
      label: "Find"
    };

    const userSearch = {
      state: {
        isLoading,
        userExists
      }
    };

    const userAccess = {
      projects,
      userName,
      userEmail
    };

    const backIcon = {
      message: "Find a user"
    };

    const dropdown = {
      items,
      selectedDropdownItem
    };

    return (
      <React.Fragment>
        <NavBar onClick={() => onLogoutClick()} />
        <div className="top-spacing" id="wrapper">
          {!isUserSearch && (
            <div className="backicon-spacing">
              <BackIcon
                backIcon={backIcon}
                onClick={() => this.onBackClick()}
              />
            </div>
          )}
          <div className="my-3 p-3 rounded shadow less-spacing-top">
            <h4 className="add-remove-text">Add or Remove User</h4>
            {isUserSearch && (
              <UserSearch
                userSearch={userSearch}
                inputField={inputField}
                onChange={e => this.onInputChange(e)}
                generalButton={generalButton}
                onClick={() => this.onButtonClick()}
                onKeyEnter={e => this.onKeyEnter(e)}
              />
            )}

            {!isUserSearch && (
              <UserAccess
                userAccess={userAccess}
                onXClick={id => this.removeUserFromProject(id)}
                onPlusClick={() => this.addUserToProject()}
                onDropdownClick={item => this.updateSelectedDropdownItem(item)}
                dropdown={dropdown}
                duplicateErrorMessage={duplicateErrorMessage}
              />
            )}
          </div>
        </div>
      </React.Fragment>
    );
  }
}

MainPage.propTypes = {
  onLogoutClick: PropTypes.func.isRequired
};
