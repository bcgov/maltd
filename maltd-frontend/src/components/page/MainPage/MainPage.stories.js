/* eslint-disable react/jsx-filename-extension */
import React from "react";
import { storiesOf } from "@storybook/react";
import "bootstrap/dist/css/bootstrap.css";
import MainPage from "./MainPage";

const userSearch = {
  state: { isLoading: false }
};

const isLoadingUserSearch = {
  state: { isLoading: true }
};

const inputField = {
  name: "idir",
  type: "text",
  placeholder: "Enter IDIR username to find",
  valid: false,
  invalid: false,
  value: "",
  disabled: false
};

const generalButton = {
  type: "submit",
  color: "primary",
  disabled: true,
  label: "Find"
};

storiesOf("MainPage", module)
  .add("default", () => (
    <MainPage
      inputField={inputField}
      generalButton={generalButton}
      userSearch={userSearch}
    />
  ))
  .add("valid", () => (
    <MainPage
      userSearch={userSearch}
      inputField={{ ...inputField, valid: true, value: "IDIR/validuser" }}
      generalButton={{ ...generalButton, disabled: false }}
    />
  ))
  .add("loading", () => (
    <MainPage
      userSearch={isLoadingUserSearch}
      inputField={{
        ...inputField,
        valid: true,
        value: "IDIR/validuser",
        disabled: true
      }}
      generalButton={generalButton}
    />
  ))
  .add("invalid", () => (
    <MainPage
      userSearch={userSearch}
      inputField={{ ...inputField, invalid: true, value: "a" }}
      generalButton={{ ...generalButton, color: "danger" }}
    />
  ));
