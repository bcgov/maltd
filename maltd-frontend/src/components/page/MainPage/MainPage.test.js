import React from "react";
import renderer from "react-test-renderer";
import axios from "axios";
import Adapter from "enzyme-adapter-react-16";
import Enzyme, { shallow } from "enzyme";
import MockAdapter from "axios-mock-adapter";
import waitUntil from "async-wait-until";
import MainPage from "./MainPage";

Enzyme.configure({ adapter: new Adapter() });

describe("Main page", () => {
  test("Component renders as expected", () => {
    const component = renderer.create(<MainPage />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  let wrapper;
  let instance;
  let mock;
  const baseUrl = process.env.REACT_APP_MALTD_API
    ? process.env.REACT_APP_MALTD_API
    : "https://localhost:5001";

  beforeEach(() => {
    wrapper = shallow(<MainPage />);
    instance = wrapper.instance();
    mock = new MockAdapter(axios);
  });

  describe("onBackClick", () => {
    test("Function modifies local state and takes user to user search screen as expected", () => {
      const clearFormFunc = jest.spyOn(MainPage.prototype, "clearForm");

      wrapper.setState({ isUserSearch: false });
      wrapper.find("BackIcon").simulate("click");

      expect(instance.state.isUserSearch).toBe(true);
      expect(clearFormFunc).toHaveBeenCalled();
      expect(instance.state.userExists).toBe(null);
      expect(instance.state.value).toBe("");
      expect(instance.state.isLoading).toBe(false);
      expect(instance.state.disabledButton).toBe(true);
      expect(instance.state.disabledInput).toBe(false);
      expect(instance.state.invalidInput).toBe(false);
      expect(instance.state.validInput).toBe(false);
    });
  });

  describe("updateSelectedDropdownItem", () => {
    test("Function updates the selected dropdown item", () => {
      const selectedProject = { id: "1", name: "project", type: "type" };

      wrapper.setState({ isUserSearch: false });
      wrapper
        .find("UserAccess")
        .props()
        .onDropdownClick(selectedProject);

      expect(instance.state.selectedDropdownItem).toBe(selectedProject);
    });
  });

  describe("onInputChange", () => {
    test("Function updates the input field with empty input", () => {
      const event = { target: { value: "" } };

      wrapper
        .find("UserSearch")
        .props()
        .onChange(event);

      expect(instance.state.userExists).toBe(null);
      expect(instance.state.invalidInput).toBe(false);
      expect(instance.state.validInput).toBe(false);
      expect(instance.state.disabledButton).toBe(true);
      expect(instance.state.color).toBe("primary");
      expect(instance.state.value).toBe(event.target.value);
    });

    test("Function updates the input field with input less than 3 characters in length", () => {
      const event = { target: { value: "aa" } };

      wrapper
        .find("UserSearch")
        .props()
        .onChange(event);

      expect(instance.state.userExists).toBe(null);
      expect(instance.state.invalidInput).toBe(true);
      expect(instance.state.color).toBe("danger");
      expect(instance.state.value).toBe(event.target.value);
    });

    test("Function updates the input field with valid input", () => {
      const event = { target: { value: "validinput" } };

      wrapper
        .find("UserSearch")
        .props()
        .onChange(event);

      expect(instance.state.userExists).toBe(null);
      expect(instance.state.invalidInput).toBe(false);
      expect(instance.state.validInput).toBe(true);
      expect(instance.state.disabledButton).toBe(false);
      expect(instance.state.color).toBe("primary");
      expect(instance.state.value).toBe(event.target.value);
    });
  });

  describe("onButtonClick", () => {
    test("Catches error when /api/projects endpoint is not called successfully", () => {
      const clearFormFunc = jest.spyOn(MainPage.prototype, "clearForm");
      const data = { response: true };
      mock.onGet(`${baseUrl}/api/projects`).reply(400, data);

      wrapper
        .find("UserSearch")
        .props()
        .onClick()
        .catch(() => {
          expect(clearFormFunc).toHaveBeenCalled();
        });
    });
  });

  describe("removeUserFromProject", () => {
    test("Function is called when x icon is clicked", () => {
      const removeUserFromProject = jest.spyOn(
        MainPage.prototype,
        "removeUserFromProject"
      );
      const id = 123;

      wrapper.setState({ isUserSearch: false });
      wrapper
        .find("UserAccess")
        .props()
        .onXClick(id);

      expect(removeUserFromProject).toHaveBeenCalled();
    });
  });

  describe("addUserToProject", () => {
    test("Function is called when plus icon is clicked", () => {
      const addUserToProject = jest.spyOn(
        MainPage.prototype,
        "addUserToProject"
      );
      wrapper.setState({ selectedDropdownItem: { id: 122 } });

      wrapper.setState({ isUserSearch: false });
      wrapper
        .find("UserAccess")
        .props()
        .onPlusClick();

      expect(addUserToProject).toHaveBeenCalled();
    });

    test("Function should make network request and should update state on success", async done => {
      const mock = new MockAdapter(axios);

      expect(wrapper.state().projects).toEqual([]);

      wrapper.setState({
        isUserSearch: false,
        selectedDropdownItem: { id: 123 },
        value: "val"
      });
      mock
        .onPut(
          `${baseUrl}/api/projects/${
            wrapper.state().selectedDropdownItem.id
          }/users/${wrapper.state().value}`
        )
        .reply(200);

      wrapper
        .find("UserAccess")
        .props()
        .onPlusClick();

      await waitUntil(() => {
        return wrapper.state().projects;
      });

      expect(wrapper.state().projects).toEqual([{ id: 123 }]);
      done();
    });
  });
});
