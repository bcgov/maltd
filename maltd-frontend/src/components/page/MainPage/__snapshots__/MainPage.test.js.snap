// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page Component renders as expected 1`] = `
<main>
  <header>
    <nav
      className="container-fluid navbar navbar-expand-lg navbar-dark"
    >
      <div
        aria-labelledby="title"
        className="navbar-brand pointer"
      />
      <img
        alt="B.C. Government Logo"
        className="img-fluid d-none d-md-block pointer"
        height="44"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="presentation"
        src="undefined/images/bcid-logo-rev-en.svg"
        width={181}
      />
      <img
        alt="B.C. Government Logo"
        className="img-fluid d-md-none pointer"
        height="44"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="presentation"
        src="undefined/images/bcid-symbol-rev.svg"
        width={64}
      />
      <div
        aria-labelledby="title"
        className="pointer navbar-brand nav-item nav-link"
      />
      <div
        className="navbar-brand"
        id="title"
      >
        MALTD
      </div>
    </nav>
  </header>
  <div
    className="navbar-styling"
  >
    <div />
    <button
      className="bcgov-button logout-btn btn"
      data-test-id=""
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Logout
    </button>
  </div>
  <div
    className="page"
  >
    <div
      className="content col-md-12"
    >
      <div
        className="my-3 p-3 rounded shadow less-spacing-top"
      >
        <h2>
          Add or Remove User
        </h2>
        <div
          className="move-down"
        >
          <p>
            Find a user by IDIR username and add them to or remove them from projects.
          </p>
          <p />
          <div
            className="inner"
          >
            <input
              autoFocus={true}
              className="input-field my-2 form-control"
              disabled={false}
              name="idir"
              onChange={[Function]}
              onKeyPress={[Function]}
              placeholder="Enter IDIR username to find"
              type="text"
              value=""
            />
            <p />
            <div
              className="error-btn-div"
            >
              <div />
              <div
                className="d-flex justify-content-end"
              >
                <button
                  className="bcgov-button normal-blue btn"
                  data-test-id=""
                  disabled={true}
                  onClick={[Function]}
                  type="button"
                >
                  Find
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <footer
    className="footer"
  >
    <nav
      className="navbar navbar-expand-sm navbar-dark justify-content-end"
    >
      <button
        className="navbar-toggler"
        data-target="#footerBar"
        data-toggle="collapse"
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="collapse navbar-collapse flex-grow-0"
        id="footerBar"
      >
        <ul
          className="navbar-nav text-right"
        >
          <li
            className="nav-item m-1"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca"
              rel="noopener noreferrer"
              target="_blank"
            >
              BC Government
            </a>
          </li>
          <li
            className="nav-item m-1"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/disclaimer"
              rel="noopener noreferrer"
              target="_blank"
            >
              Disclaimer
            </a>
          </li>
          <li
            className="nav-item m-1"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/privacy"
              rel="noopener noreferrer"
              target="_blank"
            >
              Privacy
            </a>
          </li>
          <li
            className="nav-item m-1"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/accessibility"
              rel="noopener noreferrer"
              target="_blank"
            >
              Accessibility
            </a>
          </li>
          <li
            className="nav-item m-1"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/copyright"
              rel="noopener noreferrer"
              target="_blank"
            >
              Copyright
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </footer>
</main>
`;
